lib/common/redis.c

```diff
@@ -38,21 +59,60 @@ static void on_redis_connect(const redisAsyncContext *redis, int status)
         conn->state = H2O_REDIS_CONNECTION_STATE_CLOSED;
         conn->_redis = NULL;
         if (conn->on_close != NULL) {
-            conn->on_close(redis->errstr);
+            conn->on_close(redis->c.errstr);
         }
     }
+
+    if (h2o_timeout_is_linked(&conn->_connect_timeout_entry)) {
+        h2o_timeout_unlink(&conn->_connect_timeout_entry);
+        h2o_timeout_dispose(conn->loop, &conn->_connect_timeout);
+    }
 }
 
-static void on_redis_disconnect(const redisAsyncContext *redis, int status)
+static void close_connection(h2o_redis_conn_t *conn)
 {
-    h2o_redis_conn_t *conn = (h2o_redis_conn_t *)redis->data;
     conn->state = H2O_REDIS_CONNECTION_STATE_CLOSED;
-    conn->_redis = NULL;
+
     if (conn->on_close != NULL) {
-        conn->on_close(redis->errstr);
+        conn->on_close(conn->_redis->c.errstr);
+    }
+    conn->_redis = NULL;
+
+    if (h2o_timeout_is_linked(&conn->_connect_timeout_entry)) {
+        h2o_timeout_unlink(&conn->_connect_timeout_entry);
+    }
+    if (h2o_linklist_is_linked(&conn->_connect_timeout._link)) {
+        h2o_timeout_dispose(conn->loop, &conn->_connect_timeout);
     }
 }
 
+static void on_disconnect(const redisAsyncContext *redis, int status)
```

Comment:
Does the introduction of the function mean that we now have a "disconnecting" (or "closing") state?

If yes, is there a need to add the state to `h2o_redis_connection_state_t`? If no, why do we need this function?

I would also appreciate it if you could consider unifying the term being used. At the moment I see "close" and "disconnect".

Reply:
> Does the introduction of the function mean that we now have a "disconnecting" (or "closing") state?

In our design, disconnecting hiredis object cannot be touched because once disconnecting procedure is fired our redis client immediately detach hiredis object and try to create new connection if another request comes. Regarding that, I don't think we have to have `disconnecting` state.